Route order preference (#718)
irrdnet/irrd: docs/admins/object-suppression.rst

===========================
Object suppression overview
===========================

IRRd supports three methods to suppress objects:

* :doc:`RPKI </admins/rpki>`, where IRRd suppresses objects that are
  invalid according to
  `RFC6811 origin validation <https://tools.ietf.org/html/rfc6811>`_.
  In addition, IRRd will create pseudo-IRR objects representing ROAs.
* The :doc:`scope filter </admins/scopefilter>`, which suppresses objects
  matching certain configured prefixes and/or AS numbers.
* The :doc:`route object preference </admins/route-object-preference>`,
  which suppresses objects that overlap with other route objects from
  sources with a higher preference.

> Suppression is a kind of pseudo delete/create: IRRd may act as if the

Author: ```suggestion
Suppression is a kind of pseudo delete/create: IRRd will act as if the
```

object was deleted, but it may become visible again later. Unlike normal

> deletions and creations,

Author: add note about PK conflicts

this visibility change may have an external cause, e.g. a new ROA created
by the address space holder, or a change in scope filter configuration.

> Suppression can apply even to objects for which the instance

Author: ```suggestion
Suppression can also apply to objects for which the instance
```

is not authoritative.

> Periodic updates

Author: fold these three sections together

----------------
RPKI and route object preference periodically update the status for all
relevant objects in the database. This resolves issues around temporary
inconsistencies and, for RPKI, is also the moment when the local ROA
storage is updated.

For RPKI, this happens every ``rpki.roa_import_timer``.

The scope filter updates all statuses on startup and then after any
configuration change.

Receiving updates from mirrors
------------------------------
When IRRd receives objects from mirrors, in full imports or over NRTM,

> the RPKI, scope filter and route preference status of all objects is updated.

Author: ```suggestion
the RPKI, scope filter and route preference status of these objects is updated.
```

Authoritative changes
---------------------

> For objects for which this instance is authoritative, the scope

Author: reword, too many for's

filter and RPKI will reject creation of a suppressed object.

Scope of suppression
--------------------
Suppressed objects are not visible in the responses to queries,

> database exports or NRTM streams.

Author: ```suggestion
database exports, NRTM streams, or the event stream.
```

recorded in your local journal.

Therefore, both the local state of your IRRd, and anyone mirroring from
your IRRd, will be up to date with the RPKI status.

For route object preference, there are cases where IRRd records an ADD
and then immediately a DEL for the same object.

Interaction between suppression statuses
----------------------------------------
Each suppression status is independent, but for the visibility in
the journal and query responses, IRRd considers the combined state:
an object is suppressed if any of the suppression methods see it as
suppressed.

Following on the previous example: if along with the RPKI state,
the object is out of scope according to the scope filter, IRRd will
not generate an ADD in the journal at any point, and the object
will never be visible in query responses.

Important considerations when enabling
--------------------------------------
When you enable one of these features for the first time, the periodic
task will usually take considerably longer than on subsequent runs.
On the first run, a large number of objects in the database may need to
be updated, whereas this number is much smaller on subsequent runs.

> The same may occur after changing settings, if these change a large number

Author: add to last paragraph

of objects.

While the import and validation is running, processing other
database changes may be delayed.

These large status changes may also generate a significant amount
of local journal entries, which are used to generate NRTM responses
for anyone mirroring any source from your IRRd. Depending on the

> sources, there may be in the order of 100K-1M NRTM updates.

Author: ```suggestion
sources, there may be in the order of 100K-1M journal entries.
```

> NRTM is not really designed for this scale, and therefore it is

Author: ```suggestion
NRTM is really not designed for this scale, and therefore it is
```

likely faster to have mirrors reload their copy.

Writing such huge numbers of changes to the journal also takes
considerable time.

> In order to speed up enabling, if you already want to have mirrors

Author: weird sentence
